frontend/src/app/main/ui/shapes/text.cljs

```diff
@@ -83,33 +89,70 @@
 
     base))
 
+(defn get-all-fonts [node]
+  (let [current-font (if (not (nil? (:font-id node)))
+                       #{(:font-id node)}
+                       #{})
+        children-font (map get-all-fonts (:children node))]
+    (reduce set/union (conj children-font current-font))))
+
+
+(defn fetch-font [font-id]
+  (let [{:keys [family variants]} (get @fonts/fontsdb font-id)]
+    (-> (js/fetch (fonts/gfont-url family variants))
+        (p/then (fn [res] (.text res))))))
+
+(defn embed-font [font-id]
+  (p/let [font-text (fetch-font font-id)
+          url-to-data (->> font-text
+                        (re-seq #"url\(([^)]+)\)")
+                        (map second)
+                        (map df/fetch-as-data-uri)
+                        (p/all))]
+    (reduce (fn [text [url data]] (str/replace text url data)) font-text url-to-data)))
+
 (defn- render-text-node
   ([node] (render-text-node 0 node))
   ([index {:keys [type text children] :as node}]
    (mf/html
-    (if (string? text)
-      (let [style (generate-text-styles (clj->js node))]
-        [:span {:style style :key index} text])
-      (let [children (map-indexed render-text-node children)]
-        (case type
-          "root"
-          (let [style (generate-root-styles (clj->js node))]
-            [:div.root.rich-text
-             {:key index
-              :style style
-              :xmlns "http://www.w3.org/1999/xhtml"}
-             children])
-
-          "paragraph-set"
-          (let [style #js {:display "inline-block"
-                           :width "100%"}]
-            [:div.paragraphs {:key index :style style} children])
-
-          "paragraph"
-          (let [style (generate-paragraph-styles (clj->js node))]
-            [:p {:key index :style style} children])
-
-          nil))))))
+    (let [embed-resources? (mf/use-ctx muc/embed-ctx)
```

Comment:
i think this let (and the following effects) shold be outside the `mf/html`

Reply:
OK